File: src/main/java/com/github/alexandrecarlton/idea/settings/layout/IdeaSettings.java
package com.github.alexandrecarlton.idea.settings.layout;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.github.alexandrecarlton.idea.settings.layout.editor.EditorSettings;
import com.github.alexandrecarlton.idea.settings.layout.editor.general.GeneralSettings;
import org.immutables.value.Value;

import java.util.Optional;

@Value.Immutable
@JsonDeserialize(as = ImmutableIdeaSettings.class)
public interface IdeaSettings {

  Optional<EditorSettings> editor();

  Optional<GeneralSettings> general();

}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/layout/IdeaSettings.kt
package com.github.alexandrecarlton.idea.settings.layout

import com.fasterxml.jackson.annotation.JsonProperty
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.BuildExecutionDeploymentSettings
import com.github.alexandrecarlton.idea.settings.layout.configurations.ConfigurationSettings
import com.github.alexandrecarlton.idea.settings.layout.editor.EditorSettings
import com.github.alexandrecarlton.idea.settings.layout.languages_frameworks.LanguagesFrameworksSettings
import com.github.alexandrecarlton.idea.settings.layout.other_settings.OtherSettings
import com.github.alexandrecarlton.idea.settings.layout.project_settings.ProjectSettingsSettings
import com.github.alexandrecarlton.idea.settings.layout.tools.ToolsSettings

// TODO: This should have three top level properties:
// - Project Structure
// - Settings
// - Configurations
data class IdeaSettings(

    @JsonProperty("Editor")
    val editor: EditorSettings? = null,

    @JsonProperty("Build, Execution, Deployment")
    val buildExecutionDeployment: BuildExecutionDeploymentSettings? = null,

    @JsonProperty("Languages & Frameworks")
    val languagesFrameworks: LanguagesFrameworksSettings? = null,

    @JsonProperty("Tools")
    val tools: ToolsSettings? = null,

    @JsonProperty("Other Settings")
    val otherSettings: OtherSettings? = null,

    @JsonProperty("Project Settings")
    val projectSettings: ProjectSettingsSettings? = null,

    @JsonProperty("Configurations")
    val configurations: List<ConfigurationSettings>? = null
)

